oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -97,33 +118,157 @@ private static String encodeCert(X509Certificate certificate)
-   * @return The JSON string representation of the base64-encoded leaf certificate in a JSON array.
-   * @throws IOException If an I/O error occurs while reading the certificate file.
+   * @return The JSON string representation of the base64-encoded certificate chain (leaf
+   *     certificate followed by the trust chain, if present).
+   * @throws IOException If an I/O error occurs while reading the certificate file(s).
    */
   @Override
   public String getSubjectToken(ExternalAccountSupplierContext context) throws IOException {
+    String leafCertPath = credentialSource.getCredentialLocation();
+    String trustChainPath = null;
+    if (credentialSource.getCertificateConfig() != null) {
+      trustChainPath = credentialSource.getCertificateConfig().getTrustChainPath();
+    }
+
+    // Load and encode the leaf certificate.
+    String encodedLeafCert = loadAndEncodeLeafCertificate(leafCertPath);
+
+    // Initialize the certificate chain for the subject token. The Security Token Service (STS)
+    // requires that the leaf certificate (the one used for authenticating this workload) must be
+    // the first certificate in this chain.
+    java.util.List<String> certChain = new java.util.ArrayList<>();
+    certChain.add(encodedLeafCert);
+
+    // Handle trust chain loading and processing.
     try {
-      // credentialSource.credentialLocation is expected to be non-null here,
-      // set during IdentityPoolCredentials construction for certificate type.
-      X509Certificate leafCert = loadLeafCertificate(credentialSource.getCredentialLocation());
-      String encodedLeafCert = encodeCert(leafCert);
+      // Read the trust chain.
+      List<X509Certificate> trustChainCerts = readTrustChain(trustChainPath);
+
+      // Process the trust chain certificates read from the file.
+      if (!trustChainCerts.isEmpty()) {
+        // Get the first certificate from the user-provided trust chain file.
+        X509Certificate firstTrustCert = trustChainCerts.get(0);
+        String encodedFirstTrustCert = encodeCert(firstTrustCert);
+
+        // If the first certificate in the user-provided trust chain file is *not* the leaf
+        // certificate (which has already been added as the first element to `certChain`), then add
+        // this certificate to `certChain`. This handles cases where the user's trust chain file
+        // starts with an intermediate certificate. If the first certificate in the trust chain file
+        // *is* the leaf certificate, this means the user has explicitly included the leaf in their
+        // trust chain file. In this case, we skip adding it again to prevent duplication, as the
+        // leaf is already at the beginning of `certChain`.
+        if (!encodedFirstTrustCert.equals(encodedLeafCert)) {
+          certChain.add(encodedFirstTrustCert);
+        }
+
+        // Iterate over the remaining certificates in the trust chain.
+        for (int i = 1; i < trustChainCerts.size(); i++) {
```

Comment:
nit: iterating from 1 looks a bit funny to me especially given the code duplication for 0, I think the more idiomatic way to write this for-loop is to iterate from 0, then check for i==0 inside the for-loop for the special case handling.

Reply:
Thanks for the suggestion! I see your point about the loop starting at 1 and the duplication of get()/encodeCert().

My reasoning for the current structure was to keep the logic for handling the first certificate in the trust chain (which has the rule: 'add if not leaf') distinctly separate from the logic for subsequent certificates (which have the rule: 'throw if leaf, else add'). I think this separation, aided by the comments, makes the different rules very explicit.

Combining them into a single loop with an if (i == 0) check would make the loop body a bit more branched. Given that the core conditional logic for adding/throwing is quite different for the first element versus the rest, I found the current separation slightly clearer to follow for these distinct cases.